pkg/geoipupdate/database/local_file_writer.go

```diff
@@ -0,0 +1,166 @@
+			return nil, errors.Wrap(err, "Encountered an error while createing hash for file "+filePath)
+		}
+		dbWriter.oldHash = fmt.Sprintf("%x", hash.Sum(nil))
+		if verbose {
+			log.Printf("Calculated MD5 sum for %s: %s", filePath, dbWriter.oldHash)
+		}
+	}
+	if err := dbWriter.lockDirectory(); err != nil {
+		return nil, err
+	}
+
+	dbWriter.swapFile, err = os.Create(fmt.Sprintf("%s.swap", dbWriter.filePath))
+	if err != nil {
+		return nil, errors.Wrap(err, "Error creating swap file")
+	}
+
+	return dbWriter, nil
+}
+
+func (writer *LocalFileDatabaseWriter) lockDirectory() error {
+	fi, err := os.Stat(filepath.Dir(writer.filePath))
+	if err != nil {
+		return errors.Wrap(err, "database directory is not available")
+	}
+	if !fi.IsDir() {
+		return errors.New("database directory is not a directory")
+	}
+	if writer.verbose {
+		log.Printf("Acquired lock file lock (%s)", writer.lockFile)
+	}
+	writer.lock = flock.New(writer.lockFile)
+	ok, err := writer.lock.TryLock()
+	if err != nil {
+		return errors.Wrap(err, "error acquiring a lock")
+	}
+	if !ok {
+		return errors.Errorf("could not acquire lock on %s", writer.lockFile)
+	}
+	return nil
+}
+
+//Write writes data to swap file
+func (writer *LocalFileDatabaseWriter) Write(p []byte) (n int, err error) {
+	return writer.swapFile.Write(p)
+}
+
+//Close closes the swap file
+func (writer *LocalFileDatabaseWriter) Close() (err error) {
+	_ = writer.swapFile.Close()
+	_ = os.Remove(writer.swapFile.Name())
+	if err := writer.lock.Unlock(); err != nil {
+		return errors.Wrap(err, "error releasing lock file")
+	}
+	return nil
+}
+
+//ValidHash checks that the swap file's MD5 matches the expectedHash
+func (writer *LocalFileDatabaseWriter) ValidHash(expectedHash string) error {
+	_, _ = writer.swapFile.Seek(0, 0)
+	md5Writer := md5.New()
+	reader, err := os.Open(writer.swapFile.Name())
+	defer func() {
+		if err := reader.Close(); err != nil {
+			log.Println(errors.Wrap(err, "Encountered an error closing swap file after validating its hash"))
+		}
+	}()
+	if err != nil {
+		return errors.Wrap(err, "swap file was unable to be opened")
+	}
+	if _, err := io.Copy(md5Writer, reader); err != nil {
+		return errors.Wrap(err, "no X-Database-MD5 header found")
+	}
+	hash := fmt.Sprintf("%x", md5Writer.Sum(nil))
+	if !strings.EqualFold(hash, expectedHash) {
+		return errors.Errorf("MD5 of new database (%s) does not match expected MD5 (%s)", hash, expectedHash)
+	}
+	return nil
+}
+
+//SetFileModificationTime explicitly sets the database's file write time to the provided time
+func (writer *LocalFileDatabaseWriter) SetFileModificationTime(lastModified time.Time) error {
+	if err := os.Chtimes(writer.filePath, lastModified, lastModified); err != nil {
+		return errors.Wrap(err, "error setting times on file")
+	}
+	return nil
+}
+
+//Commit renames the swap file to the name of the database file before syncing the directory
+func (writer *LocalFileDatabaseWriter) Commit() error {
+	_ = writer.swapFile.Close()
```

Comment:
Checking the error now.  Previously the Sync was removed because the MD5 was calculated on the closed file.  Per one of your other requests, the MD5 is now calculated as the file is being written to save on having to scan through multiple times.  I've added the Sync back in accordingly.